src/util/PageRange.cpp

```diff
@@ -1,76 +1,75 @@
+ * results in a vector containing the following tuples, given `pageCount = 10`:
+ *
+ *     {0, 2}, {5, 5}, {3, 9}, {0, 4}, {5, 5}, {9, 9}, {0,9}
+ *
+ * Note that the page numbers are parsed in the format 1 - pageCount (permissive) and
+ * the return values are in the range 0 - (pageCount-1).
+ */
+std::vector<PageRangeEntry> PageRange::parse(const std::string& s, size_t pageCount) {
+    const std::regex separators("[,;:]");
+    // break the string into comma (or ;:) separated tokens.
+    auto begin = std::sregex_token_iterator(s.begin(), s.end(), separators, -1);
+    auto end = std::sregex_token_iterator();
+
+    std::vector<PageRangeEntry> entries;
+    const std::regex pagerange_regex("\\s*(\\d*)\\s*(-)?\\s*(\\d*).*");
+    // process each token as a range, ignoring whitespace
+    for (auto it = begin; it != end; it++) {
+        std::smatch match;
+        PageRangeEntry entry;
+        std::string p = it->str();
+        if (std::regex_match(p, match, pagerange_regex)) {
+            std::string first = match[1].str();  // the first number (optional)
+            std::string dash = match[2].str();   // if set, we parsed a range, otherwise a sole number (in first).
+            std::string last = match[3].str();   // the second number (optional)
+            if (first.empty() && dash.empty()) {
+                continue;
+            } else if (first.empty()) {
+                entry.first = 1;
+            } else {
+                std::from_chars(first.data(), first.data() + first.size(), entry.first);
             }
-            p = next;
-        }
-
-        end = start;
-
-        while (isspace(*p)) {
-            p++;
-        }
-
-        if (*p == '-') {
-            p++;
-            end = static_cast<int>(strtol(p, &next, 10));
-            if (next == p)  // a half-open range like 2-
-            {
-                end = pageCount;
-            } else if (end < start) {
-                end = start;
+            if (dash.empty()) {
+                last = first;
             }
-        }
-
-        data.push_back(new PageRangeEntry(start - 1, end - 1));
-
-        // Skip until end or separator
-        while (*p && !isSeparator(*p)) {
-            p++;
-        }
-
-        // if not at end, skip separator
-        if (*p) {
-            p++;
+            if (last.empty()) {
+                entry.last = pageCount;
+            } else {
+                std::from_chars(last.data(), last.data() + last.size(), entry.last);
+            }
+            if (entry.first == 0)
+                entry.first = 1;
+            else if (entry.first > pageCount)
+                entry.first = pageCount;
+            if (entry.last > pageCount)
+                entry.last = pageCount;
+            if (entry.last < entry.first)
+                entry.last = entry.first;
+            if (entry.first != 0) {
+                entry.first--;
+            }
+            if (entry.last != 0) {
+                entry.last--;
+            }
```

Comment:
Please document what all these changes do. Even though it might be obvious after reading the full code, comments should be added wherever you do not understand the meaning right away.

Reply:
Thanks, I'm working on it. I discovered a bug: `pageCount=0` would still push elements into the returned vector, but it should not. I've explained the situation in comments.

As soon as I can rework some repository issue I'm having right now, I'll push the changes.

Reply:
OK comments added.